internal/pkg/docker/orchestrator/orchestrator.go

```diff
@@ -0,0 +1,328 @@
+				errCh <- fmt.Errorf("stop %q: %w", name, err)
+				return
+			}
+			errCh <- nil
+		}()
+	}
+
+	var errs []error
+	for err := range errCh {
+		errs = append(errs, err)
+		if len(errs) == len(task.Containers) {
+			break
+		}
+	}
+
+	return errors.Join(errs...)
+}
+
+// waitForContainerToStart blocks until the container specified by id starts.
+func (o *Orchestrator) waitForContainerToStart(ctx context.Context, id string) error {
+	for {
+		isRunning, err := o.docker.IsContainerRunning(ctx, id)
+		switch {
+		case err != nil:
+			return fmt.Errorf("check if %q is running: %w", id, err)
+		case isRunning:
+			return nil
+		}
+
+		select {
+		case <-time.After(1 * time.Second):
+		case <-ctx.Done():
+			return ctx.Err()
+		}
+	}
+}
+
+// containerID returns the full ID for a container with name run by s.
+func (o *Orchestrator) containerID(name string) string {
+	return o.idPrefix + name
+}
+
+// Task defines a set of Containers to be run together.
+// Containers within a Task can talk to each other on localhost
+// and are stopped and started as a group.
+type Task struct {
+	Containers map[string]ContainerDefinition
+}
+
+// ContainerDefinition defines information necessary to run a container.
+type ContainerDefinition struct {
+	ImageURI string
+	EnvVars  map[string]string
+	Secrets  map[string]string
+	Ports    map[string]string // host port -> container port
+}
+
+// pauseRunOptions returns RunOptions for the pause container for t.
+// The pause container owns the networking namespace that is shared
+// among all of the containers in the task.
+func (o *Orchestrator) pauseRunOptions(t Task) dockerengine.RunOptions {
+	opts := dockerengine.RunOptions{
+		ImageURI:       pauseContainerURI,
+		ContainerName:  o.containerID("pause"),
+		Command:        []string{"sleep", "infinity"},
+		ContainerPorts: make(map[string]string),
+	}
+
+	for _, ctr := range t.Containers {
+		for hostPort, ctrPort := range ctr.Ports {
+			// TODO some error if host port is already defined?
+			opts.ContainerPorts[hostPort] = ctrPort
+		}
+	}
+	return opts
+}
+
+// containerRunOptions returns RunOptions for the given container.
+func (o *Orchestrator) containerRunOptions(name string, ctr ContainerDefinition) dockerengine.RunOptions {
+	return dockerengine.RunOptions{
+		ImageURI:         ctr.ImageURI,
+		ContainerName:    o.containerID(name),
+		EnvVars:          ctr.EnvVars,
+		Secrets:          ctr.Secrets,
+		ContainerNetwork: o.containerID("pause"),
+		LogOptions:       o.logOptions(name, ctr),
+	}
+}
+
+// run calls `docker run` using opts. Errors are only returned
+// to the main Orchestrator routine if the taskID the container was run with
+// matches the current taskID the Orchestrator is running.
+func (o *Orchestrator) run(taskID int32, opts dockerengine.RunOptions) {
+	o.wg.Add(1)
+	go func() {
+		defer o.wg.Done()
+
+		if err := o.docker.Run(context.Background(), &opts); err != nil {
+			curTaskID := o.curTaskID.Load()
+			if curTaskID == stoppedTaskID {
```

Comment:
When would this happen 🤔 ? When the caller asks the orchestrator to `Stop()`, followed by a `RunTask()` call?

Reply:
basically the idea is that `Stop()` _causes_ `docker.Run()` to return an error _for a previous `RunTask()` call. So the error case is actually:
1. `RunTask()`
2. `Stop()`
3. we hit this logic

this is because `docker.Run()` **blocks** until the container itself stops, which happens whenever we call `docker.Stop()` in `stopTask()`. Unless the container exits on it's own with a `0`, `docker.Run()` always returns an error.

Reply:
> ⬆️ exactly why I asked

For some reason I totally misread the question 😂

Reply:
I feel like this is probably ok for now but the fact that depending on "`docker.Run()` to only allow running once at a time" sounds a little off to me 💭

Reply:
nvm i think i get it now.